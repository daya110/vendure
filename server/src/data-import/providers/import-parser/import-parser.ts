import { Injectable } from '@nestjs/common';
import * as parse from 'csv-parse';
import { Stream } from 'stream';

export interface RawProductRecord {
    name?: string;
    slug?: string;
    description?: string;
    assets?: string;
    optionGroups?: string;
    optionValues?: string;
    sku?: string;
    price?: string;
    taxCategory?: string;
    variantAssets?: string;
}

export interface ParsedProductVariant {
    optionValues: string[];
    sku: string;
    price: number;
    taxCategory: string;
    assetPaths: string[];
}

export interface ParsedProduct {
    name: string;
    slug: string;
    description: string;
    assetPaths: string[];
    optionGroups: Array<{
        name: string;
        values: string[];
    }>;
}

export interface ParsedProductWithVariants {
    product: ParsedProduct;
    variants: ParsedProductVariant[];
}

/**
 * Validates and parses CSV files into a data structure which can then be used to created new entities.
 */
@Injectable()
export class ImportParser {
    async parseProducts(input: string | Stream): Promise<ParsedProductWithVariants[]> {
        const options: parse.Options = {
            columns: true,
            trim: true,
        };
        return new Promise<ParsedProductWithVariants[]>((resolve, reject) => {
            if (typeof input === 'string') {
                parse(input, options, (err, records: RawProductRecord[]) => {
                    if (err) {
                        reject(err);
                    }

                    try {
                        const output = this.processRawRecords(records);
                        resolve(output);
                    } catch (err) {
                        reject(err);
                    }
                });
            } else {
                const parser = parse(options);
                const records: RawProductRecord[] = [];
                // input.on('open', () => input.pipe(parser));
                input.pipe(parser);
                parser.on('readable', () => {
                    let record;
                    // tslint:disable-next-line:no-conditional-assignment
                    while ((record = parser.read())) {
                        records.push(record);
                    }
                });
                parser.on('error', reject);
                parser.on('end', () => {
                    try {
                        const output = this.processRawRecords(records);
                        resolve(output);
                    } catch (err) {
                        reject(err);
                    }
                });
            }
        });
    }

    private processRawRecords(records: RawProductRecord[]): ParsedProductWithVariants[] {
        const output: ParsedProductWithVariants[] = [];
        let currentRow: ParsedProductWithVariants | undefined;
        validateColumns(records[0]);
        for (const r of records) {
            if (r.name) {
                if (currentRow) {
                    populateOptionGroupValues(currentRow);
                    output.push(currentRow);
                }
                currentRow = {
                    product: parseProductFromRecord(r),
                    variants: [parseVariantFromRecord(r)],
                };
            } else {
                if (currentRow) {
                    currentRow.variants.push(parseVariantFromRecord(r));
                }
            }
            validateOptionValueCount(r, currentRow);
        }
        if (currentRow) {
            populateOptionGroupValues(currentRow);
            output.push(currentRow);
        }
        return output;
    }
}

function populateOptionGroupValues(currentRow: ParsedProductWithVariants) {
    const values = currentRow.variants.map(v => v.optionValues);
    currentRow.product.optionGroups.forEach((og, i) => {
        const uniqueValues = Array.from(new Set(values.map(v => v[i])));
        og.values = uniqueValues;
    });
}

function validateColumns(r: RawProductRecord) {
    const requiredColumns: Array<keyof RawProductRecord> = [
        'name',
        'slug',
        'description',
        'assets',
        'optionGroups',
        'optionValues',
        'sku',
        'price',
        'taxCategory',
        'variantAssets',
    ];
    const rowKeys = Object.keys(r);
    const missing: string[] = [];
    for (const col of requiredColumns) {
        if (!rowKeys.includes(col)) {
            missing.push(col);
        }
    }
    if (missing.length) {
        throw new Error(
            `The import file is missing the following columns: ${missing.map(m => `"${m}"`).join(', ')}`,
        );
    }
}

function validateOptionValueCount(r: RawProductRecord, currentRow?: ParsedProductWithVariants) {
    if (!currentRow) {
        return;
    }
    const optionValues = parseStringArray(r.optionValues);
    if (currentRow.product.optionGroups.length !== optionValues.length) {
        throw new Error(
            `The number of optionValues must match the number of optionGroups for the product "${r.name}"`,
        );
    }
}

function parseProductFromRecord(r: RawProductRecord): ParsedProduct {
    return {
        name: parseString(r.name),
        slug: parseString(r.slug),
        description: parseString(r.description),
        assetPaths: parseStringArray(r.assets),
        optionGroups: parseStringArray(r.optionGroups).map(name => ({
            name,
            values: [],
        })),
    };
}

function parseVariantFromRecord(r: RawProductRecord): ParsedProductVariant {
    return {
        optionValues: parseStringArray(r.optionValues),
        sku: parseString(r.sku),
        price: parseNumber(r.price),
        taxCategory: parseString(r.taxCategory),
        assetPaths: parseStringArray(r.variantAssets),
    };
}

function parseString(input?: string): string {
    return (input || '').trim();
}

function parseNumber(input?: string): number {
    return +(input || '').trim();
}

function parseStringArray(input?: string, separator = ','): string[] {
    return (input || '')
        .trim()
        .split(separator)
        .map(s => s.trim())
        .filter(s => s !== '');
}
