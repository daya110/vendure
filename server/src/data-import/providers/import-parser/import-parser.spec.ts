import * as fs from 'fs-extra';
import * as path from 'path';

import { ImportParser } from './import-parser';

describe('ImportParser', () => {
    describe('parseProducts', () => {
        it('single product with a single variant', async () => {
            const importParser = new ImportParser();

            const input = await loadTestFixture('single-product-single-variant.csv');
            const result = await importParser.parseProducts(input);

            expect(result).toMatchSnapshot();
        });

        it('single product with a multiple variants', async () => {
            const importParser = new ImportParser();

            const input = await loadTestFixture('single-product-multiple-variants.csv');
            const result = await importParser.parseProducts(input);

            expect(result).toMatchSnapshot();
        });

        it('multiple products with multiple variants', async () => {
            const importParser = new ImportParser();

            const input = await loadTestFixture('multiple-products-multiple-variants.csv');
            const result = await importParser.parseProducts(input);

            expect(result).toMatchSnapshot();
        });

        it('works with streamed input', async () => {
            const importParser = new ImportParser();

            const filename = path.join(__dirname, 'test-fixtures', 'multiple-products-multiple-variants.csv');
            const input = fs.createReadStream(filename);
            const result = await importParser.parseProducts(input);

            expect(result).toMatchSnapshot();
        });

        describe('error conditions', () => {
            it('throws on invalid option values', async () => {
                const importParser = new ImportParser();

                const input = await loadTestFixture('invalid-option-values.csv');
                try {
                    await importParser.parseProducts(input);
                    fail('should have thrown');
                } catch (err) {
                    expect(err.message).toBe(
                        'The number of optionValues must match the number of optionGroups for the product "Artists Smock"',
                    );
                }
            });

            it('throws on ivalid columns', async () => {
                const importParser = new ImportParser();

                const input = await loadTestFixture('invalid-columns.csv');
                try {
                    await importParser.parseProducts(input);
                    fail('should have thrown');
                } catch (err) {
                    expect(err.message).toBe(
                        'The import file is missing the following columns: "slug", "assets"',
                    );
                }
            });
        });
    });
});

function loadTestFixture(fileName: string): Promise<string> {
    return fs.readFile(path.join(__dirname, 'test-fixtures', fileName), 'utf-8');
}
