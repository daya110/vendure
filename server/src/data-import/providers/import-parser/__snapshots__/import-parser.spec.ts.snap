// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportParser parseProducts multiple products with multiple variants 1`] = `
Array [
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "A great device for stretching paper.",
      "name": "Perfect Paper Stretcher",
      "optionGroups": Array [
        Object {
          "name": "size",
          "values": Array [
            "Half Imperial",
            "Quarter Imperial",
            "Full Imperial",
          ],
        },
      ],
      "slug": "Perfect-paper-stretcher",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Half Imperial",
        ],
        "price": 45.3,
        "sku": "PPS12",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Quarter Imperial",
        ],
        "price": 32.5,
        "sku": "PPS14",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Full Imperial",
        ],
        "price": 59.5,
        "sku": "PPSF",
        "taxCategory": "standard",
      },
    ],
  },
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "Mabef description",
      "name": "Mabef M/02 Studio Easel",
      "optionGroups": Array [],
      "slug": "mabef-m02-studio-easel",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [],
        "price": 910.7,
        "sku": "M02",
        "taxCategory": "standard",
      },
    ],
  },
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "Really mega pencils",
      "name": "Giotto Mega Pencils",
      "optionGroups": Array [
        Object {
          "name": "box size",
          "values": Array [
            "Box of 8",
            "Box of 12",
          ],
        },
      ],
      "slug": "giotto-mega-pencils",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Box of 8",
        ],
        "price": 4.16,
        "sku": "225400",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Box of 12",
        ],
        "price": 6.24,
        "sku": "225600",
        "taxCategory": "standard",
      },
    ],
  },
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "Keeps the paint off the clothes",
      "name": "Artists Smock",
      "optionGroups": Array [
        Object {
          "name": "size",
          "values": Array [
            "small",
            "large",
          ],
        },
        Object {
          "name": "colour",
          "values": Array [
            "beige",
            "navy",
          ],
        },
      ],
      "slug": "artists-smock",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "small",
          "beige",
        ],
        "price": 11.99,
        "sku": "10112",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "large",
          "beige",
        ],
        "price": 11.99,
        "sku": "10113",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "small",
          "navy",
        ],
        "price": 11.99,
        "sku": "10114",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "large",
          "navy",
        ],
        "price": 11.99,
        "sku": "10115",
        "taxCategory": "standard",
      },
    ],
  },
]
`;

exports[`ImportParser parseProducts single product with a multiple variants 1`] = `
Array [
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "A great device for stretching paper.",
      "name": "Perfect Paper Stretcher",
      "optionGroups": Array [
        Object {
          "name": "size",
          "values": Array [
            "Half Imperial",
            "Quarter Imperial",
            "Full Imperial",
          ],
        },
      ],
      "slug": "perfect-paper-stretcher",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Half Imperial",
        ],
        "price": 45.3,
        "sku": "PPS12",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Quarter Imperial",
        ],
        "price": 32.5,
        "sku": "PPS14",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Full Imperial",
        ],
        "price": 59.5,
        "sku": "PPSF",
        "taxCategory": "standard",
      },
    ],
  },
]
`;

exports[`ImportParser parseProducts single product with a single variant 1`] = `
Array [
  Object {
    "product": Object {
      "assetPaths": Array [
        "pps1.jpg",
        "pps2.jpg",
      ],
      "description": "A great device for stretching paper.",
      "name": "Perfect Paper Stretcher",
      "optionGroups": Array [],
      "slug": "perfect-paper-stretcher",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [],
        "price": 45.3,
        "sku": "PPS12",
        "taxCategory": "standard",
      },
    ],
  },
]
`;

exports[`ImportParser parseProducts works with streamed input 1`] = `
Array [
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "A great device for stretching paper.",
      "name": "Perfect Paper Stretcher",
      "optionGroups": Array [
        Object {
          "name": "size",
          "values": Array [
            "Half Imperial",
            "Quarter Imperial",
            "Full Imperial",
          ],
        },
      ],
      "slug": "Perfect-paper-stretcher",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Half Imperial",
        ],
        "price": 45.3,
        "sku": "PPS12",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Quarter Imperial",
        ],
        "price": 32.5,
        "sku": "PPS14",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Full Imperial",
        ],
        "price": 59.5,
        "sku": "PPSF",
        "taxCategory": "standard",
      },
    ],
  },
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "Mabef description",
      "name": "Mabef M/02 Studio Easel",
      "optionGroups": Array [],
      "slug": "mabef-m02-studio-easel",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [],
        "price": 910.7,
        "sku": "M02",
        "taxCategory": "standard",
      },
    ],
  },
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "Really mega pencils",
      "name": "Giotto Mega Pencils",
      "optionGroups": Array [
        Object {
          "name": "box size",
          "values": Array [
            "Box of 8",
            "Box of 12",
          ],
        },
      ],
      "slug": "giotto-mega-pencils",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Box of 8",
        ],
        "price": 4.16,
        "sku": "225400",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "Box of 12",
        ],
        "price": 6.24,
        "sku": "225600",
        "taxCategory": "standard",
      },
    ],
  },
  Object {
    "product": Object {
      "assetPaths": Array [],
      "description": "Keeps the paint off the clothes",
      "name": "Artists Smock",
      "optionGroups": Array [
        Object {
          "name": "size",
          "values": Array [
            "small",
            "large",
          ],
        },
        Object {
          "name": "colour",
          "values": Array [
            "beige",
            "navy",
          ],
        },
      ],
      "slug": "artists-smock",
    },
    "variants": Array [
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "small",
          "beige",
        ],
        "price": 11.99,
        "sku": "10112",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "large",
          "beige",
        ],
        "price": 11.99,
        "sku": "10113",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "small",
          "navy",
        ],
        "price": 11.99,
        "sku": "10114",
        "taxCategory": "standard",
      },
      Object {
        "assetPaths": Array [],
        "optionValues": Array [
          "large",
          "navy",
        ],
        "price": 11.99,
        "sku": "10115",
        "taxCategory": "standard",
      },
    ],
  },
]
`;
